amcl:
  ros__parameters:
    alpha1: 0.2
    alpha2: 0.2
    alpha3: 0.2
    alpha4: 0.2
    alpha5: 0.2
    base_frame_id: "base_footprint"
    beam_skip_distance: 0.5
    beam_skip_error_threshold: 0.9
    beam_skip_threshold: 0.3
    do_beamskip: false
    global_frame_id: "map"
    lambda_short: 0.1
    laser_likelihood_max_dist: 1.2
    laser_max_range: 8.0
    laser_min_range: 0.15
    laser_model_type: "likelihood_field"
    max_beams: 60
    max_particles: 2000
    min_particles: 500
    odom_frame_id: "odom"
    pf_err: 0.05
    pf_z: 0.99
    recovery_alpha_fast: 0.0
    recovery_alpha_slow: 0.0
    resample_interval: 1
    robot_model_type: "nav2_amcl::DifferentialMotionModel"
    save_pose_rate: 0.5
    sigma_hit: 0.2
    tf_broadcast: true
    transform_tolerance: 1.0
    update_min_a: 0.2
    update_min_d: 0.25
    z_hit: 0.5
    z_max: 0.05
    z_rand: 0.25
    z_short: 0.2
    scan_topic: /scan

bt_navigator:
  ros__parameters:
    transform_tolerance: 0.5
    global_frame: map
    robot_base_frame: base_footprint
    odom_topic: odom
    bt_loop_duration: 10
    filter_duration: 0.3
    default_server_timeout: 30
    wait_for_service_timeout: 1000
    navigators: ["navigate_to_pose", "navigate_through_poses"]
    navigate_to_pose:
      plugin: "nav2_bt_navigator::NavigateToPoseNavigator"
      # enable_groot_monitoring: false
      # groot_server_port: 1667
    navigate_through_poses:
      plugin: "nav2_bt_navigator::NavigateThroughPosesNavigator"
      # enable_groot_monitoring: false
      # groot_server_port: 1669
    # bt_search_directories:
      # - $(find-pkg-share nav2_bt_navigator)/behavior_trees
    # 'default_nav_through_poses_bt_xml' and 'default_nav_to_pose_bt_xml' are use defaults:
    # nav2_bt_navigator/navigate_to_pose_w_replanning_and_recovery.xml
    # nav2_bt_navigator/navigate_through_poses_w_replanning_and_recovery.xml
    # They can be set here or via a RewrittenYaml remap from a parent launch file to Nav2.

    # plugin_lib_names is used to add custom BT plugins to the executor (vector of strings).
    # Built-in plugins are added automatically
    # plugin_lib_names: []

    error_code_name_prefixes:
      - assisted_teleop
      - backup
      - compute_path
      - drive_on_heading
      - follow_path
      - nav_thru_poses
      - nav_to_pose
      - route
      - spin
      - smoother
      - wait

controller_server:
  ros__parameters:
    transform_tolerance: 0.5
    controller_frequency: 10.0  # fixed: 20.0
    costmap_update_timeout: 0.30
    min_x_velocity_threshold: 0.02  # fixed: 0.001 -> 0.6
    min_y_velocity_threshold: 0.0  # fixed: 0.001 -> 0.0001
    min_theta_velocity_threshold: 0.02
    failure_tolerance: 0.5  # fixed: 0.3
    progress_checker_plugins: ["progress_checker"]
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    current_progress_checker: "progress_checker"
    current_goal_checker: "general_goal_checker"
    controller_plugins: ["FollowPath"]
    use_realtime_priority: false
    speed_limit_topic: "speed_limit"

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.08  # fixed: 0.5
      movement_time_allowance: 8.0  # fixed: 10.0 -> 15.0
    # Goal checker parameters
    #precise_goal_checker:
    #  plugin: "nav2_controller::SimpleGoalChecker"
    #  xy_goal_tolerance: 0.25
    #  yaw_goal_tolerance: 0.25
    #  stateful: true
    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.25
      yaw_goal_tolerance: 0.25
    FollowPath:
      plugin: "nav2_mppi_controller::MPPIController"
      time_steps: 40  # fixed: 56
      model_dt: 0.1  # fixed: 0.05
      batch_size: 1000  # fixed: 1000
      ax_max: 1.5
      ax_min: -1.5
      az_max: 3.5
      vx_std: 0.8  # fixed: 0.2
      vy_std: 0.0
      wz_std: 0.4
      vx_max: 0.4
      vx_min: -0.4
      wz_max: 0.6
      # wz_min: 0.3
      iteration_count: 1
      prune_distance: 1.3
      transform_tolerance: 0.5  # fixed: 0.1
      temperature: 0.5  # fixed: 0.3
      gamma: 0.02  # fixed: 0.015
      motion_model: "DiffDrive"
      visualize: true
      publish_optimal_trajectory: true
      regenerate_noises: true
      TrajectoryVisualizer:
        trajectory_step: 5
        time_step: 3
      TrajectoryValidator:
        # The validator can be configured with additional parameters if needed.
        plugin: "mppi::DefaultOptimalTrajectoryValidator"
        collision_lookahead_time: 2.0  # Time to look ahead to check for collisions
        consider_footprint: false  # Whether to consider the full robot's footprint (or circle) in trajectory validation collision checks
        min_vel_x: -1.5
        max_vel_x: 1.5
        max_vel_theta: 1.9
        max_acc_x: 3.0
        max_acc_theta: 3.5
      # AckermannConstraints:
        # min_turning_r: 0.2
      critics:
        [
          "ConstraintCritic",
          "CostCritic",
          "GoalCritic",
          "GoalAngleCritic",
          "PathAlignCritic",
          "PathFollowCritic",
          "PathAngleCritic",
          "PreferForwardCritic"
        ]
      ConstraintCritic:
        enabled: true
        cost_power: 1
        cost_weight: 4.0
      GoalCritic:
        enabled: true
        cost_power: 1
        cost_weight: 5.0
        threshold_to_consider: 1.4
      GoalAngleCritic:
        enabled: true
        cost_power: 1
        cost_weight: 3.0
        threshold_to_consider: 0.5
      PreferForwardCritic:
        enabled: true
        cost_power: 1
        cost_weight: 8.0 # fixed: 5.0 -> 15.0
        threshold_to_consider: 0.5
      CostCritic:
        enabled: true
        cost_power: 1
        cost_weight: 4.0  # fixed: 1.6
        # tweak cost_weight if obstacle avoidance feels jittery or unsafe
        near_collision_cost: 253
        critical_cost: 300.0
        consider_footprint: false
        collision_cost: 1000000.0
        near_goal_distance: 1.0
        trajectory_point_step: 2
      PathAlignCritic:
        enabled: true
        cost_power: 1
        cost_weight: 10.0  # fixed: 14.0 -> 6.0
        max_path_occupancy_ratio: 0.30
        trajectory_point_step: 4
        threshold_to_consider: 0.5
        offset_from_furthest: 20
        use_path_orientations: false  # change "true" if robot does weird motion
      PathFollowCritic:
        enabled: true
        cost_power: 1
        cost_weight: 15.0
        offset_from_furthest: 5
        threshold_to_consider: 1.4
      PathAngleCritic:
        enabled: true
        cost_power: 1
        cost_weight: 5.0
        offset_from_furthest: 4
        threshold_to_consider: 0.5
        max_angle_to_furthest: 1.0
        mode: 0
        # change "1" if the robot does weird motion
        # change "2" if the robot is in tight corridors
      TwirlingCritic:
      # erase # if the robot turns at the same spot
        enabled: true
        twirling_cost_power: 1
        twirling_cost_weight: 10.0

local_costmap:
  local_costmap:
    ros__parameters:
      transform_tolerance: 0.5
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: odom
      robot_base_frame: base_footprint
      rolling_window: true
      width: 5  # fixed: 3.0
      height: 5  # fixed: 3.0
      resolution: 0.05
      robot_radius: 0.44
      plugins: ["obstacle_layer", "inflation_layer"]
      # filters: ["keepout_filter"]
      # keepout_filter:
        # plugin: "nav2_costmap_2d::KeepoutFilter"
        # enabled: KEEPOUT_ZONE_ENABLED
        # filter_info_topic: "keepout_costmap_filter_info"
        # override_lethal_cost: true
        # lethal_override_cost: 200
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 5.0  # fixed: 3.0 -> 6.0 -> 8.0
        inflation_radius: 0.3  # fixed: 0.70 -> 0.5 -> 0.8-> 0.4 -> 0.15
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: true
        observation_sources: scan_merged scan
        scan_merged:        
          topic: /scan_merged
          max_obstacle_height: 2.0
          clearing: true
          marking: true
          data_type: "LaserScan"
          raytrace_max_range: 3.3
          raytrace_min_range: 0.0
          obstacle_max_range: 3.3  # fixed: 2.5
          obstacle_min_range: 0.05
          observation_persistence: 0.05
          inf_is_valid: true
          # expected_update_rate: 0.20
        scan:
          topic: /scan
          data_type: LaserScan
          sensor_frame: lidar_frame
          marking: true
          clearing: true               # ← LiDAR는 clearing 허용 (빈 공간을 멀리까지 비움)
          obstacle_range: 3.5
          raytrace_range: 6.5
          inf_is_valid: true
      # static_layer:
        # plugin: "nav2_costmap_2d::StaticLayer"
        # map_subscribe_transient_local: true
      always_send_full_costmap: false  # fixed: true

global_costmap:
  global_costmap:
    ros__parameters:
      transform_tolerance: 0.5
      update_frequency: 1.0
      publish_frequency: 1.0
      global_frame: map
      robot_base_frame: base_footprint
      robot_radius: 0.44
      resolution: 0.05
      track_unknown_space: true
      plugins: ["static_layer", "obstacle_layer", "inflation_layer"]
      # filters: ["keepout_filter", "speed_filter"]
      # keepout_filter:
        # plugin: "nav2_costmap_2d::KeepoutFilter"
        # enabled: KEEPOUT_ZONE_ENABLED
        # filter_info_topic: "keepout_costmap_filter_info"
        # override_lethal_cost: True
        # lethal_override_cost: 200
      # speed_filter:
        # plugin: "nav2_costmap_2d::SpeedFilter"
        # enabled: SPEED_ZONE_ENABLED
        # filter_info_topic: "speed_costmap_filter_info"
        # speed_limit_topic: "speed_limit"
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: true
        observation_sources: scan_merged scan_raw
        scan_merged:
          # A relative topic will be appended to the parent of the global_costmap namespace.
          # For example:
          #   * User chosen namespace is `tb4`.
          #   * User chosen topic is `scan`.
          #   * Topic will be remapped to `/tb4/scan` without `global_costmap`.
          #   * Use global topic `/scan` if you do not wish the node namespace to apply
          topic: /scan_merged
          max_obstacle_height: 2.0
          clearing: false
          marking: true
          data_type: "LaserScan"
          raytrace_max_range: 3.3
          raytrace_min_range: 0.2
          obstacle_max_range: 3.3
          obstacle_min_range: 0.2
          expected_update_rate: 0.20
          observation_persistence: 0.2
        scan_raw:
          topic: /scan
          data_type: "LaserScan"
          marking: false
          clearing: true
          obstacle_max_range: 8.0
          raytrace_max_range: 10.0
          obstacle_min_range: 0.1
          raytrace_min_range: 0.1
          expected_update_rate: 0.5
          observation_persistence: 0.5
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: true
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 8.0  # fixed: 3.0
        inflation_radius: 0.3  # fixed: 0.7 -> 0.8 -> 0.4 -> 0.15
      always_send_full_costmap: false  # fixed: true

# The yaml_filename does not need to be specified since it going to be set by defaults in launch.
# If you'd rather set it in the yaml, remove the default "map" value in the launch file(s).
# file & provide full path to map below. If CLI map configuration or launch default is provided, that will be used.
map_server:
  ros__parameters:
    yaml_filename: "/home/st/map_sensor10.yaml"

# keepout_filter_mask_server:
  # ros__parameters:
    # topic_name: "keepout_filter_mask"
    # yaml_filename: ""

# keepout_costmap_filter_info_server:
  # ros__parameters:
    # type: 0
    # filter_info_topic: "keepout_costmap_filter_info"
    # mask_topic: "keepout_filter_mask"
    # base: 0.0
    # multiplier: 1.0

# speed_filter_mask_server:
  # ros__parameters:
    # topic_name: "speed_filter_mask"
    # yaml_filename: ""

# speed_costmap_filter_info_server:
  # ros__parameters:
    # type: 1
    # filter_info_topic: "speed_costmap_filter_info"
    # mask_topic: "speed_filter_mask"
    # base: 100.0
    # multiplier: -1.0

map_saver:
  ros__parameters:
    save_map_timeout: 5.0
    free_thresh_default: 0.25
    occupied_thresh_default: 0.65
    map_subscribe_transient_local: true

planner_server:
  ros__parameters:
    transform_tolerance: 0.5
    expected_planner_frequency: 5.0  # fixed: 20.0
    planner_plugins: ["GridBased"]
    costmap_update_timeout: 1.0
    GridBased:
      plugin: "nav2_navfn_planner::NavfnPlanner"
      tolerance: 0.5
      use_astar: true
      lethal_cost: 253
      neutral_cost: 66
      cost_factor: 0.6
      allow_unknown: false

smoother_server:
  ros__parameters:
    smoother_plugins: ["route_smoother"]
    route_smoother:
      plugin: "nav2_smoother::SimpleSmoother"
      tolerance: 1.0e-10
      max_its: 1000
      refinement_num: 5
      # False for Route Server or others where smoothing out sharp corners is OK
      do_refinement: true

behavior_server:
  ros__parameters:
    local_costmap_topic: local_costmap/costmap_raw
    global_costmap_topic: global_costmap/costmap_raw
    local_footprint_topic: local_costmap/published_footprint
    global_footprint_topic: global_costmap/published_footprint
    cycle_frequency: 10.0
    behavior_plugins:
      ["spin", "backup", "drive_on_heading", "assisted_teleop", "wait"]
    spin:
      plugin: "nav2_behaviors::Spin"
    backup:
      plugin: "nav2_behaviors::BackUp"
      # acceleration_limit: 2.5
      # deceleration_limit: -2.5
      # minimum_speed: 0.10
    drive_on_heading:
      plugin: "nav2_behaviors::DriveOnHeading"
      # acceleration_limit: 2.5
      # deceleration_limit: -2.5
      # minimum_speed: 0.10
    wait:
      plugin: "nav2_behaviors::Wait"
    assisted_teleop:
      plugin: "nav2_behaviors::AssistedTeleop"
    local_frame: odom
    global_frame: map
    robot_base_frame: base_footprint
    transform_tolerance: 0.5
    simulate_ahead_time: 2.0
    max_rotational_vel: 0.6
    min_rotational_vel: 0.4
    rotational_acc_lim: 3.2
    enable_stamped_cmd_vel: false

waypoint_follower:
  ros__parameters:
    loop_rate: 20
    stop_on_failure: false
    waypoint_task_executor_plugin: "wait_at_waypoint"
    wait_at_waypoint:
      plugin: "nav2_waypoint_follower::WaitAtWaypoint"
      enabled: True
      waypoint_pause_duration: 200  # if CPU occurs problem, then change it 2000

velocity_smoother:
  ros__parameters:
    smoothing_frequency: 20.0
    scale_velocities: False
    feedback: "OPEN_LOOP"
    max_velocity: [0.4, 0.0, 0.6]
    min_velocity: [-0.6, 0.0, -0.6]
    max_accel: [0.8, 0.0, 1.1]  # [2.5, 0.0, 3.2]
    max_decel: [-0.8, 0.0, -1.1] # [-2.5, 0.0, -3.2]
    odom_topic: "odom"
    odom_duration: 0.1
    deadband_velocity: [0.05, 0.0, 0.05]
    velocity_timeout: 1.0
    enable_stamped_cmd_vel: false

collision_monitor:
  ros__parameters:
    base_frame_id: "base_footprint"
    odom_frame_id: "odom"
    cmd_vel_in_topic: "cmd_vel_smoothed"
    cmd_vel_out_topic: "cmd_vel"
    state_topic: "collision_monitor_state"
    transform_tolerance: 0.5
    source_timeout: 1.0
    base_shift_correction: true
    stop_pub_timeout: 2.0
    # Polygons represent zone around the robot for "stop", "slowdown" and "limit" action types,
    # and robot footprint for "approach" action type.
    polygons: ["FootprintApproach", "PolygonStop", "PolygonSlow"]
    FootprintApproach:
      type: "polygon"
      action_type: "approach"
      footprint_topic: "local_costmap/published_footprint"
      time_before_collision: 1.2
      simulation_time_step: 0.1
      min_points: 6
      visualize: true
      enabled: true
    PolygonStop: 
      type: "circle"
      radius: 0.50
      action_type: "stop"
      min_points: 5
      visualize: true
      enabled: true
      stop_pub_timeout: 0.5
    PolygonSlow:
      type: "circle"
      radius: 0.70
      action_type: "slowdown"
      min_points: 4
      slowdown_ratio: 0.5
      visualize: true
      enabled: true
    observation_sources: ["merged_slow"]
    merged_slow:
      type: scan
      topic: /scan_merged
      min_range: 0.05
      max_range: 3.0
      enabled: true
